Refactored run factor and added unit test

DaveOkpare opened the pull request:
## Describe your changes

## Issue ticket number and link
Closes #161 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [x] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

neelasha23 commented on src/sql/run.py:
```diff
@@ -381,39 +381,50 @@ def _commit(conn, config):
-            return resultset.DataFrame()
-        elif config.autopolars:
-            return resultset.PolarsDataFrame()
+    if not sql.strip():
```
I think we should add a comment here in which cases this condition will be true.

neelasha23 commented on src/sql/run.py:
```diff
@@ -381,39 +381,50 @@ def _commit(conn, config):
+    for statement in sqlparse.split(sql):
+        first_word = sql.strip().split()[0].lower()
+        if first_word == "begin":
+            raise Exception("ipython_sql does not support transactions")
```
change this to custom exception

neelasha23 commented on src/sql/run.py:
```diff
@@ -381,39 +381,50 @@ def _commit(conn, config):
+            print(interpret_rowcount(result.rowcount))
+
+    resultset = ResultSet(result, config)
+    if config.autopandas:
```
this can be converted to one function

neelasha23 commented on src/tests/test_run.py:
```diff
@@ -0,0 +1,39 @@
+
+
+@pytest.mark.parametrize(
+    "sql, config, user_namespace",
```
Why do we need `parametrize` here?

neelasha23 commented on src/tests/test_run.py:
```diff
@@ -0,0 +1,39 @@
+        run(conn, sql, config, user_namespace)
+
+
+@pytest.mark.parametrize(
```
Again, please use `parametrize` if multiple input sets are passed

neelasha23 commented on src/sql/run.py:
```diff
@@ -381,39 +381,50 @@ def _commit(conn, config):
+    if config.autopandas:
+        return resultset.DataFrame()
+    elif config.autopolars:
+        return resultset.PolarsDataFrame()
```
please add testcases for autopolars and autopandas enabled as well

edublancas commented:
@DaveOkpare: note that we just merged a PR that modified this function so you need to rebase to fix the merging connflicts

edublancas commented:
@DaveOkpare  please address @neelasha23's comments

DaveOkpare commented:
@neelasha23 Resolved the comments

neelasha23 commented on src/tests/test_run.py:
```diff
@@ -0,0 +1,129 @@
+    )
+
+
+@mock.patch("sql.run.ResultSet")
```
Please use monkeypatching instead of patch

DaveOkpare replied:
Fixed @neelasha23

neelasha23 replied:
I still see @mock.patch used everywhere
Please refer [monkeypatch](https://docs.pytest.org/en/7.1.x/how-to/monkeypatch.html)

neelasha23 commented on src/sql/run.py:
```diff
@@ -381,39 +381,50 @@ def _commit(conn, config):
             pass  # not all engines can commit
 
 
+def is_postgres(dialect):
```
Please click on the `Resolve Conversation` button below for all the comments that have been addressed.

neelasha23 commented:
Please squash the commits. You can find some guideline [here](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md)

DaveOkpare commented:
> Please squash the commits. You can find some guideline [here](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md)

Fixed @neelasha23

DaveOkpare reviewed (commented):
@neelasha23 Please refer to this commit

neelasha23 commented:
> Please squash the commits. You can find some guideline [here](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md)

@DaveOkpare

tonykploomber commented on src/sql/run.py:
```diff
@@ -385,54 +385,82 @@ def _commit(conn, config, manual_commit):
+        first_word = sql.strip().split()[0].lower()
+        manual_commit = False
+        if first_word == "begin":
+            raise ValueError("ipython_sql does not support transactions")
```
Are we able to write the test case to cover this ValueError?

tonykploomber reviewed (changes requested):
Overall LGTM! GJ

edublancas commented:
please rebase, we recently made a change where we updated some print statements and turned into logger.debug, so should be an easy fix

DaveOkpare commented:
> please rebase, we recently made a change where we updated some print statements and turned into logger.debug, so should be an easy fix

I've squashed my commit, and rebased with your changes

neelasha23 commented:
Sure go ahead @DaveOkpare
